sound/soc/sof/core.c

```diff
@@ -455,10 +455,19 @@ EXPORT_SYMBOL(snd_sof_device_remove);
 int snd_sof_device_shutdown(struct device *dev)
 {
 	struct snd_sof_dev *sdev = dev_get_drvdata(dev);
+	struct snd_sof_pdata *pdata = sdev->pdata;
 
 	if (IS_ENABLED(CONFIG_SND_SOC_SOF_PROBE_WORK_QUEUE))
 		cancel_work_sync(&sdev->probe_work);
 
+	/*
+	 * make sure clients and machine driver(s) are unregisters to force
+	 * all userspace devices to be closed prior to the DSP shutdown sequence
+	 */
+	sof_unregister_clients(sdev);
+
+	snd_sof_machine_unregister(sdev, pdata);
```

Comment:
@plbossart if we're going to stop and close everything, do we really need to care about continuing further to anything below at all? The shutdown callback was mainly added to get the DSP out of D0i3 and if everything is closed with the unregister above, we don't really need it right?

Reply:
@ranj063 The recommendation (also in ACPI spec) is to put device into similar state as in S3. So that would mean D3 in our state so we should run sof_suspend() still.

Reply:
> @plbossart if we're going to stop and close everything, do we really need to care about continuing further to anything below at all? The shutdown callback was mainly added to get the DSP out of D0i3 and if everything is closed with the unregister above, we don't really need it right?

interesting remark @ranj063, indeed the logic is particularly convoluted here. I guess there's no real harm in trying to exit D0i3 again in case the unregister part had any sort of error? In other words that would be a 2-level safety procedure to make sure the DSP can reboot in all cases. belt and braces/suspenders.

Reply:
is there a way we can do this also on panic? We see cases where DSP is left in bad state after a panic

Reply:
@cujomalainey how will the SOF driver know that there has been a panic?

Reply:
Are there any device hooks that are called on a panic? I actually don't know, just floating ideas for another problem

Reply:
@cujomalainey I don;t think there're any hooks to guarantee the shutdown callbacks are invoked even in the case of panics. I wonder if the BIOS can do something about this? ie put the DSP in D3 during shutdown. @plbossart would you know?

Reply:
@cujomalainey @ranj063 I remember having a brilliant idea to force the shutdown to happen earlier, but with the winter break I don't recall much. 

I think the general idea was to register a reboot handler which as you can see below would be called *before* the actual device_shutdown. It's a bit silly and unlikely to be accepted upstream, but in case another driver oopses in their .shutdown() callback we could make sure the hardware is in the right state before the reboot

````c
static void kernel_shutdown_prepare(enum system_states state)
{
    blocking_notifier_call_chain(&reboot_notifier_list,
        (state == SYSTEM_HALT) ? SYS_HALT : SYS_POWER_OFF, NULL);
    system_state = state;
    usermodehelper_disable();
    device_shutdown();
}
````

Completely untested, just a brain fart at this point.

Reply:
Yea, would be good to investigate, we have seen cases where after an oops the device requires another reboot because the DSP was left is a bad state and wont boot as a result

Reply:
@cujomalainey Btw, there have been a few fixes to the DSP reboot after panics like "ASoC: SOF: pm: Force DSP off on suspend in BOOT_FAILED state also". In general runtime PM should always reboot the DSP, i.e. if user-space releases PCMs (when it hits errors), PCM goes to runtime-suspend, upon resume there should be a clean boot.

Runtime suspend is not ideal, but it's fairly safe towards user-space (user-space sees error, closes the PCM, DSP can be rebooted without existing streams). Unregister clients essentially leaves system without any PCMs, so some additional logic to reprobe the driver is needed. Similar challenges are with reboot "under the hood" (hiding DSP reboot as a big XRUN event to user-space).